CFUValidator/CFUValidator.cs

```diff
@@ -56,183 +56,102 @@ public class CFUValidator
-            ReadCheckForUpdatesXML(out xenServerPatches, out xenServerVersions, out xenCenterVersions);
+            statusReporter($"Getting check for updates XML from {_xmlLocation}...");
+            ReadCheckForUpdatesXML(out var xenServerPatches, out var xenServerVersions, out var clientVersions);
 
-            List<string> versionToCheck = GetVersionToCheck(xenServerVersions);
-
-            foreach (string ver in versionToCheck)
+            List<string> versionsToCheck;
+            if (_serverVersion == CFUCommandLineOptionManager.AllVersions)
             {
-                ServerVersion = ver;
-                RunTestsForGivenServerVersion(xenServerVersions, xenServerPatches, xenCenterVersions);
+                versionsToCheck = xenServerVersions.ConvertAll(i => i.Version.ToString()).Distinct().ToList();
+            }
+            else
+            {
+                CheckVersionExistsInCfu(_serverVersion, xenServerVersions);
+                versionsToCheck = new List<string> {_serverVersion};
             }
-            
-        }
 
+            var summaryGenerators = new List<ISummaryGenerator>();
+            foreach (string ver in versionsToCheck)
+                summaryGenerators.AddRange(RunTestsForGivenServerVersion(ver, xenServerVersions, xenServerPatches, clientVersions, statusReporter));
 
-        private List<string> GetVersionToCheck(List<XenServerVersion> xenServerVersions)
-        {
-            if(ServerVersion == CFUCommandLineOptionManager.AllVersions)
-                return xenServerVersions.ConvertAll(i => i.Version.ToString()).Distinct().ToList();
-            
-            return new List<string>{ServerVersion};
+            summaryGenerators.ForEach(s => statusReporter(s.GenerateSummary()));
         }
 
-        private void RunTestsForGivenServerVersion(List<XenServerVersion> xenServerVersions, 
-                                                   List<XenServerPatch> xenServerPatches,
-                                                   List<ClientVersion> xenCenterVersions)
+        private List<ISummaryGenerator> RunTestsForGivenServerVersion(string serverVersion, List<XenServerVersion> xenServerVersions,
+            List<XenServerPatch> xenServerPatches, List<ClientVersion> clientVersions, Action<string> statusReporter)
         {
-            CheckProvidedVersionNumber(xenServerVersions);
-            
-            Status = String.Format("Generating server {0} mock-ups...", ServerVersion);
-            SetupMocks(xenServerPatches, xenServerVersions);
+            statusReporter($"Generating server {serverVersion} mock-ups...");
+            SetupMocks(serverVersion, xenServerPatches, xenServerVersions);
 
-            Status = "Determining XenCenter update required...";
-            var xcupdateAlerts = XenAdmin.Core.Updates.NewClientUpdateAlerts(xenCenterVersions, new Version(ServerVersion));
+            statusReporter("Determining client update required...");

```

Comment:
Nit: I'd say: `"Determining if client update is required..."`.

Reply:
Hmm, judging by the output I think it's probably "Determining required client update". I corrected it: https://github.com/xenserver/xenadmin/compare/690460dca93bd4a61742fb35d83aa76003376c7e..246f3d142b1dba8c192bfc1a9048485f82760c58